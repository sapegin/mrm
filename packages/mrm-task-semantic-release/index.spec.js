jest.mock('fs');
jest.mock('git-username');
jest.mock('mrm-core/src/util/log', () => ({
	added: jest.fn(),
}));
jest.mock('mrm-core/src/npm', () => ({
	uninstall: jest.fn(),
}));

const { uninstall } = require('mrm-core');
const { getTaskOptions } = require('mrm');
const vol = require('memfs').vol;
const task = require('./index');

const console$log = console.log;

const stringify = o => JSON.stringify(o, null, '  ');

const packageJson = stringify({
	name: 'unicorn',
	scripts: {
		'semantic-release': 'semantic-release',
	},
});
const readmeMd = '# Unicorn';

beforeEach(() => {
	console.log = jest.fn();
});

afterEach(() => {
	vol.reset();
	uninstall.mockClear();
	console.log = console$log;
});

it('should add semantic-release', async () => {
	vol.fromJSON({
		'/package.json': packageJson,
		'/Readme.md': readmeMd,
	});

	task(await getTaskOptions(task, false, {}));

	expect(vol.toJSON()).toMatchSnapshot();

	// Repeat with existing Yaml workflow file
	vol.fromJSON({
		'/package.json': packageJson,
		'/Readme.md': readmeMd,
	});

	task(await getTaskOptions(task, false, {}));

	expect(vol.toJSON()).toMatchSnapshot();
});

it('should remove custom config from package.json', async () => {
	vol.fromJSON({
		'/package.json': stringify({
			name: 'unicorn',
			scripts: {
				'semantic-release': 'semantic-release',
			},
			release: {
				analyzeCommits: 'semantic-release-tamia/analyzeCommits',
			},
		}),
	});

	task(await getTaskOptions(task, false, {}));

	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
});

it('should remove semantic-release runner from Travis CI config', async () => {
	vol.fromJSON({
		'/.travis.yml': `language: node_js
node_js:
  - 8
after_success:
  - bash <(curl -s https://codecov.io/bash)
  - npm run semantic-release
  - npx semantic-release
`,
		'/package.json': packageJson,
	});

	task(await getTaskOptions(task, false, {}));

	expect(vol.toJSON()['/.travis.yml']).toMatchSnapshot();
	expect(uninstall).toBeCalledWith(['semantic-release', 'travis-deploy-once']);
});

it('should avoid checking semantic-release runner from string `after_success` Travis CI config', async () => {
	vol.fromJSON({
		'/.travis.yml': `language: node_js
node_js:
  - 8
after_success: npm run semantic-release
`,
		'/package.json': packageJson,
	});

	task(await getTaskOptions(task, false, {}));

	expect(vol.toJSON()['/.travis.yml']).toMatchSnapshot();
	expect(uninstall).toBeCalledWith(['semantic-release', 'travis-deploy-once']);
});
