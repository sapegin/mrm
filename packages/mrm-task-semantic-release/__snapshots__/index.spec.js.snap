// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add semantic-release 1`] = `
Object {
  "/.github/workflows/release.yml": "name: Semantic Release
on:
  push:
    branches:
      - master
jobs:
  release:
    name: Release
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: lts/*
      - run: npm ci
      - env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
          NPM_TOKEN: \${{ secrets.NPM_TOKEN }}
        run: npx semantic-release
",
  "/Readme.md": "# Unicorn

[![npm](https://img.shields.io/npm/v/unicorn.svg)](https://www.npmjs.com/package/unicorn)",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {}
}",
}
`;

exports[`should add semantic-release 2`] = `
Object {
  "/.github/workflows/release.yml": "name: Semantic Release
on:
  push:
    branches:
      - master
jobs:
  release:
    name: Release
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: lts/*
      - run: npm ci
      - env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
          NPM_TOKEN: \${{ secrets.NPM_TOKEN }}
        run: npx semantic-release
",
  "/Readme.md": "# Unicorn

[![npm](https://img.shields.io/npm/v/unicorn.svg)](https://www.npmjs.com/package/unicorn)",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {}
}",
}
`;

exports[`should remove custom config from package.json 1`] = `
"{
  \\"name\\": \\"unicorn\\",
  \\"scripts\\": {}
}"
`;

exports[`should remove semantic-release runner from Travis CI config 1`] = `
"language: node_js
node_js:
  - 8
after_success:
  - bash <(curl -s https://codecov.io/bash)
"
`;

exports[`should avoid checking semantic-release runner from string \`after_success\` Travis CI config 1`] = `
"language: node_js
node_js:
  - 8
after_success: npm run semantic-release
"
`;
